### PR TITLE
planner: introduce an individual physical plan for indexNestedLoopHashJoin

### PR DESCRIPTION
**Pull request by XuHuaiyu**
I'll split this PR into small ones:
- [x] add more test util JSON files https://github.com/pingcap/tidb/pull/12247
- [x] tiny code refine for tryToConstructIndexJoin https://github.com/pingcap/tidb/pull/12254
- [ ] introduce the individual physical plan for indexHashJoin

### What problem does this PR solve? 
Introduce an individual physical plan for IndexNestedLoopHashJoin.

### What is changed and how it works?
PhysicalIndexHashJoin will be constructed when `len(prop.Item) == 0` is true.

### Check List 

Tests 

Exists tests.

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

N/A

Release note

N/A

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -415,6 +414,32 @@ func (p *LogicalJoin) constructIndexMergeJoin(
...
+func (p *LogicalJoin) constructIndexHashJoin(
+	prop *property.PhysicalProperty,
+	outerIdx int,
+	innerTask task,
+	ranges []*ranger.Range,
+	keyOff2IdxOff []int,
+	path *accessPath,
+	compareFilters *ColWithCmpFuncManager,
+) []PhysicalPlan {
+	// TODO(xuhuaiyu): support keep outer order for indexHashJoin.
```
Comment (lamxTyler) [inline planner/core/exhaust_physical_plans.go]: Do we need it? I thought in this case index merge join would always performs better.

Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: Some cases need to keep order but index merge join can't make it. For example:

We need to join tables `t1` and `t2` on `t1.c = t2.c`, `t1` is the outer table. But it requires to keep order with `t1.a, t1.b`. The index merge join can only make sure of keeping `t1.c` order.

Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: That's the reason why we also need index lookup join now.

File: planner/core/task.go [hunk 2]
```diff
@@ -240,6 +240,72 @@ func (p *PhysicalIndexMergeJoin) GetCost(outerTask, innerTask task) float64 {
...
+	numPairs := outerCnt * innerCnt
+	if p.JoinType == SemiJoin || p.JoinType == AntiSemiJoin ||
+		p.JoinType == LeftOuterSemiJoin || p.JoinType == AntiLeftOuterSemiJoin {
+		if len(p.OtherConditions) > 0 {
+			numPairs *= 0.5
+		} else {
+			numPairs = 0
+		}
+	}
+	probeCost := numPairs * CPUFactor / concurrency
```
Comment (lzmhhh123) [inline planner/core/task.go]: Could this `probeCost` be directly divided by `concurrency`? If data is very much, the index hash join would not fall back to linear execution?

**Pull request merged**